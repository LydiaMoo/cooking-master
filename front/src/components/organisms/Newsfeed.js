import React, { Component } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import RecipePreview from '../molecules/RecipePreview.js'; 
import FullRecipe from '../molecules/FullRecipe.js'; 
import recipes from '../../assets/recipes.json'; 

const Container = styled.article`
	display: flex;
	flex-direction: row; 
	flex-wrap: wrap; 
	background: #FFFFFF; 
	/* border: 2px solid #635155;
	border-radius: 5px;  */
	width: 100%; 
`; 

export default class Newsfeed extends Component {

	static propTypes = {
		filter: PropTypes.oneOf(['all', 'starred']),
		search: PropTypes.string
	}

	static defaultProps = {
		filter: 'all',
		search: '',
	}

	state = {
		openedRecipe: null, 
		starred: [], 
	}
	//lifecycle methods

	componentWillReceiveProps(nextProps) {
		this.toggleCollapse(this.state.openedRecipe);
	}

	getFilteredRecipes = () => recipes
		.filter(recipe =>
			this.props.filter === 'all' ||
			this.starred(recipe)
		)
		.filter(this.textFilter);
	
	textFilter = (recipe) => this.props.search === '' || 
	recipe.name.match(new RegExp(this.props.search, 'gi')) ||
	recipe.mainIngredients.some(ingredient => ingredient.match(new RegExp(this.props.search, 'gi'))) ||
	recipe.ingredients.some(({name}) => name.match(new RegExp(this.props.search, 'gi')));

	starred = recipe => this.state.starred.some(item => item === recipe);

	toggleCollapse = (key) => {
		this.setState({openedRecipe: key});
	}

	toggleStarred = (key) => {
		(this.state.starred.find(item => item === key) ? 
		(this.setState({starred: this.state.starred.filter(i => i !== key)})):
		(this.setState({starred: this.state.starred.concat([key])})));
    ; 
  }
		
	render() {
		const display = (
			<div>
				<Container>
					{(this.state.openedRecipe !== null) ? (Object.entries(this.getFilteredRecipes()).map(([key, recipe]) => 
						(key === this.state.openedRecipe) ?
							(<FullRecipe
								onClick={() => this.toggleCollapse(null)}
								key={key}
								name={recipe.name} 
								cookingTime={recipe.cookingTime}
								ingredients={recipe.ingredients} 
								image={recipe.image}
								starOnChange={() => this.toggleStarred(recipe)}
								starred={this.starred(recipe)}
							/>
						):(null))) 
						 : ( (Object.entries(this.getFilteredRecipes()).map(([key, recipe]) => 
							(<RecipePreview
								onClick={() => this.toggleCollapse(key)}
								key={key}
								name={recipe.name} 
								image={recipe.image}
								starOnChange={() => this.toggleStarred(recipe)}
								starred={this.starred(recipe)}
							/>
						))))} 
				</Container>   
			</div>
			)
		return (
		<div>
			{display}
		</div>
		)
	}
}
