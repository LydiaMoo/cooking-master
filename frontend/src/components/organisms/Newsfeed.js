import React, { Component } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import RecipePreview from '../molecules/RecipePreview.js'; 
import FullRecipe from '../molecules/FullRecipe.js'; 
import recipes from '../../assets/recipes.json'; 

const Container = styled.article`
	display: flex;
	flex-direction: row; 
	flex-wrap: wrap; 
	background: #FFFFFF; 
	/* border: 2px solid #635155;
	border-radius: 5px;  */
	width: 100%; 
`; 

export default class Newsfeed extends Component {

	static propTypes = {
		filter: PropTypes.oneOf(['all', 'starred']),
		search: PropTypes.string
	}

	static defaultProps = {
		filter: 'all',
		search: '',
	}

	state = {
		openedRecipe: null, 
		starred: [], 
	}
	//lifecycle methods
	componentWillReceiveProps(nextProps) {
		this.handleCollapseClick(this.state.openedRecipe);
	}

	getFilteredRecipes = () => recipes
		.filter(recipe =>
			this.props.filter === 'all' ||
			this.state.starred.some(item => item === recipe)
		)
		.filter(recipe =>
			this.props.search === '' || 
			recipe.name.match(new RegExp(this.props.search, 'gi')) ||
			recipe.mainIngredients.some(ingredient => ingredient.match(new RegExp(this.props.search, 'gi'))) ||
			recipe.ingredients.some(({name}) => name.match(new RegExp(this.props.search, 'gi')))
		);

	starred = recipe => this.state.starred.some(item => item === recipe);

	handleExpandClick = (key) => {
		this.setState({openedRecipe: key});
	}

	handleCollapseClick = (key) => {
		this.setState({openedRecipe: null});
	}

	toggleStarred = (key) => {
		(this.state.starred.find(item => item === key) ? 
		(this.setState({starred: this.state.starred.filter(i => i !== key)})):
		(this.setState({starred: this.state.starred.concat([key])})));
    ; 
  }
		
	render() {
		const display = (
			<div>
				<Container>
					{Object.entries(this.getFilteredRecipes()).map(([key, recipe]) => (
						key === this.state.openedRecipe ? (
							<FullRecipe
								onClick={() => this.handleCollapseClick(key)}
								key={key}
								name={recipe.name} 
								cookingTime={recipe.cookingTime}
								ingredients={recipe.ingredients} 
								image={recipe.image}
								starOnChange={() => this.toggleStarred(recipe)}
								starred={this.starred(recipe)}
							/>
						) : (
							<RecipePreview
								onClick={() => this.handleExpandClick(key)}
								key={key}
								name={recipe.name} 
								image={recipe.image}
								starOnChange={() => this.toggleStarred(recipe)}
								starred={this.starred(recipe)}
							/>
						) 
					))} 
				</Container>   
			</div>
			)
		return (
		<div>
			{display}
		</div>
		)
	}
}
