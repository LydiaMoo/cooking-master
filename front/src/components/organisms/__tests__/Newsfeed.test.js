import React from 'react';
import renderer from 'react-test-renderer';
import Newsfeed from '../Newsfeed';
// import { shallow } from 'enzyme';
// import sinon from 'sinon';

describe('<Newsfeed />', () => {
  it('should match default snapshot', () => {
    const tree = renderer.create(<Newsfeed />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should match snapshot when filtering by full recipe name', () => {
    const tree = renderer.create(<Newsfeed search="Beef Stroganoff" />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should match snapshot when filtering by match of recipe name', () => {
    const tree = renderer.create(<Newsfeed search="Strogan" />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should match snapshot when filtering by mainIngredient of recipe', () => {
    const tree = renderer.create(<Newsfeed search="Mushrooms" />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should match snapshot when filtering by an ingredient name', () => {
    const tree = renderer.create(<Newsfeed search="Chicken Breasts" />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should match snapshot when filtering by All recipes', () => {
    const tree = renderer.create(<Newsfeed filter="all" />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should match snapshot when filtering by starred recipes', () => {
    const tree = renderer.create(<Newsfeed filter="starred" />).toJSON();
    expect(tree).toMatchSnapshot();
  });

});