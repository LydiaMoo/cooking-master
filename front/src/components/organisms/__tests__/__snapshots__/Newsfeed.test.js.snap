// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Newsfeed /> should match default snapshot 1`] = `
<div>
  <div>
    <article
      className="sc-dnqmqq cVyUAK"
    >
      <article
        className="sc-bxivhb eCmjqu"
      >
        <ul
          className="sc-ifAKCX bbyWFJ"
        >
          <li>
            <h1>
              Lemon Chicken
            </h1>
          </li>
        </ul>
        <img
          alt="Lemon Chicken"
          className="sc-EHOje iAIPz"
          src="http://assets.simplyrecipes.com/wp-content/uploads/2010/03/lemon-chicken-horiz-a-1600.jpg"
        />
        <button
          className="sc-bdVaJa iLpejz"
          onClick={[Function]}
        >
          <div
            alt="not a favorite Recipe"
            className="sc-htpNat elORsY"
          />
        </button>
        <button
          onClick={[Function]}
        >
           Click to expand 
        </button>
      </article>
      <article
        className="sc-bxivhb eCmjqu"
      >
        <ul
          className="sc-ifAKCX bbyWFJ"
        >
          <li>
            <h1>
              Beef Stroganoff
            </h1>
          </li>
        </ul>
        <img
          alt="Beef Stroganoff"
          className="sc-EHOje iAIPz"
          src="http://www.daringgourmet.com/wp-content/uploads/2014/12/Slow-Cooker-Beef-Stroganoff-2-web-edited.jpg"
        />
        <button
          className="sc-bdVaJa iLpejz"
          onClick={[Function]}
        >
          <div
            alt="not a favorite Recipe"
            className="sc-htpNat elORsY"
          />
        </button>
        <button
          onClick={[Function]}
        >
           Click to expand 
        </button>
      </article>
      <article
        className="sc-bxivhb eCmjqu"
      >
        <ul
          className="sc-ifAKCX bbyWFJ"
        >
          <li>
            <h1>
              Ceasar Salad
            </h1>
          </li>
        </ul>
        <img
          alt="Ceasar Salad"
          className="sc-EHOje iAIPz"
          src="https://www.bbcgoodfood.com/sites/default/files/styles/recipe/public/recipe_images/recipe-image-legacy-id--327831_11.jpg?itok=8GjuBS4F"
        />
        <button
          className="sc-bdVaJa iLpejz"
          onClick={[Function]}
        >
          <div
            alt="not a favorite Recipe"
            className="sc-htpNat elORsY"
          />
        </button>
        <button
          onClick={[Function]}
        >
           Click to expand 
        </button>
      </article>
      <article
        className="sc-bxivhb eCmjqu"
      >
        <ul
          className="sc-ifAKCX bbyWFJ"
        >
          <li>
            <h1>
              Chicken Ceasar Salad
            </h1>
          </li>
        </ul>
        <img
          alt="Chicken Ceasar Salad"
          className="sc-EHOje iAIPz"
          src="https://www.bbcgoodfood.com/sites/default/files/styles/recipe/public/recipe_images/recipe-image-legacy-id--327831_11.jpg?itok=8GjuBS4F"
        />
        <button
          className="sc-bdVaJa iLpejz"
          onClick={[Function]}
        >
          <div
            alt="not a favorite Recipe"
            className="sc-htpNat elORsY"
          />
        </button>
        <button
          onClick={[Function]}
        >
           Click to expand 
        </button>
      </article>
    </article>
  </div>
</div>
`;

exports[`<Newsfeed /> should match snapshot when filtering by an ingredient name 1`] = `
<div>
  <div>
    <article
      className="sc-dnqmqq cVyUAK"
    >
      <article
        className="sc-bxivhb eCmjqu"
      >
        <ul
          className="sc-ifAKCX bbyWFJ"
        >
          <li>
            <h1>
              Lemon Chicken
            </h1>
          </li>
        </ul>
        <img
          alt="Lemon Chicken"
          className="sc-EHOje iAIPz"
          src="http://assets.simplyrecipes.com/wp-content/uploads/2010/03/lemon-chicken-horiz-a-1600.jpg"
        />
        <button
          className="sc-bdVaJa iLpejz"
          onClick={[Function]}
        >
          <div
            alt="not a favorite Recipe"
            className="sc-htpNat elORsY"
          />
        </button>
        <button
          onClick={[Function]}
        >
           Click to expand 
        </button>
      </article>
      <article
        className="sc-bxivhb eCmjqu"
      >
        <ul
          className="sc-ifAKCX bbyWFJ"
        >
          <li>
            <h1>
              Chicken Ceasar Salad
            </h1>
          </li>
        </ul>
        <img
          alt="Chicken Ceasar Salad"
          className="sc-EHOje iAIPz"
          src="https://www.bbcgoodfood.com/sites/default/files/styles/recipe/public/recipe_images/recipe-image-legacy-id--327831_11.jpg?itok=8GjuBS4F"
        />
        <button
          className="sc-bdVaJa iLpejz"
          onClick={[Function]}
        >
          <div
            alt="not a favorite Recipe"
            className="sc-htpNat elORsY"
          />
        </button>
        <button
          onClick={[Function]}
        >
           Click to expand 
        </button>
      </article>
    </article>
  </div>
</div>
`;

exports[`<Newsfeed /> should match snapshot when filtering by full recipe name 1`] = `
<div>
  <div>
    <article
      className="sc-dnqmqq cVyUAK"
    >
      <article
        className="sc-bxivhb eCmjqu"
      >
        <ul
          className="sc-ifAKCX bbyWFJ"
        >
          <li>
            <h1>
              Beef Stroganoff
            </h1>
          </li>
        </ul>
        <img
          alt="Beef Stroganoff"
          className="sc-EHOje iAIPz"
          src="http://www.daringgourmet.com/wp-content/uploads/2014/12/Slow-Cooker-Beef-Stroganoff-2-web-edited.jpg"
        />
        <button
          className="sc-bdVaJa iLpejz"
          onClick={[Function]}
        >
          <div
            alt="not a favorite Recipe"
            className="sc-htpNat elORsY"
          />
        </button>
        <button
          onClick={[Function]}
        >
           Click to expand 
        </button>
      </article>
    </article>
  </div>
</div>
`;

exports[`<Newsfeed /> should match snapshot when filtering by mainIngredient of recipe 1`] = `
<div>
  <div>
    <article
      className="sc-dnqmqq cVyUAK"
    >
      <article
        className="sc-bxivhb eCmjqu"
      >
        <ul
          className="sc-ifAKCX bbyWFJ"
        >
          <li>
            <h1>
              Beef Stroganoff
            </h1>
          </li>
        </ul>
        <img
          alt="Beef Stroganoff"
          className="sc-EHOje iAIPz"
          src="http://www.daringgourmet.com/wp-content/uploads/2014/12/Slow-Cooker-Beef-Stroganoff-2-web-edited.jpg"
        />
        <button
          className="sc-bdVaJa iLpejz"
          onClick={[Function]}
        >
          <div
            alt="not a favorite Recipe"
            className="sc-htpNat elORsY"
          />
        </button>
        <button
          onClick={[Function]}
        >
           Click to expand 
        </button>
      </article>
    </article>
  </div>
</div>
`;

exports[`<Newsfeed /> should match snapshot when filtering by match of recipe name 1`] = `
<div>
  <div>
    <article
      className="sc-dnqmqq cVyUAK"
    >
      <article
        className="sc-bxivhb eCmjqu"
      >
        <ul
          className="sc-ifAKCX bbyWFJ"
        >
          <li>
            <h1>
              Beef Stroganoff
            </h1>
          </li>
        </ul>
        <img
          alt="Beef Stroganoff"
          className="sc-EHOje iAIPz"
          src="http://www.daringgourmet.com/wp-content/uploads/2014/12/Slow-Cooker-Beef-Stroganoff-2-web-edited.jpg"
        />
        <button
          className="sc-bdVaJa iLpejz"
          onClick={[Function]}
        >
          <div
            alt="not a favorite Recipe"
            className="sc-htpNat elORsY"
          />
        </button>
        <button
          onClick={[Function]}
        >
           Click to expand 
        </button>
      </article>
    </article>
  </div>
</div>
`;
